keps/sig-node/2400-node-swap/README.md

```diff
@@ -441,6 +472,113 @@ Additionally, end user may decide to disable swap completely for a Pod or a cont
+will be used in order to prevent the volumes' pages from swapping to disk.
+
+Bear in mind that the tmpfs noswap option is fairly new and is supported in kernel versions >= 6.4. However, different
+Linux distributions can decide to backport this options to older versions of the kernel. Therefore, when
+`--fail-swap-on=false` is being provided on a node:
+* If the kernel version equals or is above 6.4, the tmpfs noswap option is being used when necessary.
+* Else, kubelet would try to mount a dummy volume with the tmpfs noswap option to understand whether the option is
+  backported. If the mount succeeds, the tmpfs noswap option is being used when necessary.
+* Else, kubelet would raise a warning about the option not being supported and the possible risk.
+
+In the longer term, when this option would be very widely supported, this would no longer be a concern, hence this logic
+could be dropped.
+
+#### Evictions
+
+Since this KEP aims to revolve around basic swap enablement and leave more advanced features to
+follow-up KEPs, swap-based evictions are out-of-scope for this KEP
+(see the [summary](#summary) section above for more info).
+
+##### Brief technical overview of swap and evictions
+
+With swap being disabled, the memory eviction threshold should be set in a way that will let kubelet's
+eviction manager identify memory spikes in time so it can execute custom logic and take care of reclaiming
+memory and evicting pods before the kernel's OOM killer is invoked.
+
+However, with swap enabled, the situation is fundamentally different.
+In order for the kernel to start swapping memory, either the pod must breach its memory limits
+(that are not relevant for kubelet-level evictions) or the node must use all of its physical memory.
+Since swapping is a heavy operation in terms of performance, the kernel will try avoiding
+it as long as it can, and will start swapping only when it lacks physical memory.
+
+Therefore, when swap is enabled, it is acceptable to use all the available RAM which will result in the node
+using some of the swap memory.
+
+##### Current eviction limitations
+
+Because evictions are out-of-scope for this KEP, the eviction manager would remain swap-unaware.
+
+Let's say that on a specific node, the kernel would start swapping when there is less than X free physical memory
+available (see more detailed explanation [below](#advanced-best-practices-for-manually-setting-memory-evictions)).
+This means either one of the two behaviors will take place:
+1. Memory evictions threshold > X: In this scenario, kubelet's eviction manager would kick in before the kernel has
+   a chance to swap memory.
+   This means that the kernel would generally not be able to swap memory (unless the memory spike
+   is fast enough for the eviction manager to handle).
+1. Memory evictions threshold <= X: In this scenario, the kernel would start swapping memory before the memory threshold 
```

Comment:
What's the expected behavior after the swap capacity has been exhausted? Do we expect the memory eviction to kick in after that?

Reply:
If swap capacity is exhausted in the rare chance, oomkiller would step in. The main issue we see right now is that we are too aggressive with our memory eviction so pods are being evicted due to hitting the memory limit.

IMO this is probably the best approach we can do given the current design of Swap. Some of the user requests have been to effectively treat virtual memory = memory + swap. And allow for nodes to have a extra memory capacity and accept the limitations that swap is much lower than RAM.

Reply:
in the "Memory evictions threshold <= swap level" case, we actually would evict before pods can swap, thus basically invalidating swap being set
in the "Memory evictions threshold > swap level" case, it may not be fully deterministic. either i'd expect the oom killer to kick in, or evictions, once the swap limit was met. It probably depends on node memory pressure

Reply:
> in the "Memory evictions threshold > swap level" case, it may not be fully deterministic. either i'd expect the oom killer to kick in, or evictions, once the swap limit was met. It probably depends on node memory pressure

Hmm.. @haircommander the non-deterministic behavior makes it harder to expect what will happen when the node is under memory pressure. Is there still any benefit of enabling memory eviction in your opinion? 

> However, evictions will happen as soon as you are using more memory than physically present on the node. This even applies when parts of your used memory has been swapped.

> Still I personally would recommend to disable memory evictions when using the feature.

@jabdoa2 this is a good example. I'm trying to understand if this feature as-it-is is basically incompatible with the kubelet-driven eviction. And is it okay to simply ask users to disable the eviction, or is it better to make basic enhancement of the eviction to make it swap-aware.

Reply:
I personally see a good argument that we advocate users disable memory eviction if swap is enabled, with the caveat that we intend on improving swappy eviction in the future. How do you feel about that @yujuhong ?

Reply:
This will be a big breaking change if we disable eviction once swap is enabled. This will keep the KEP in Beta from my perspective. Also should we disable it for swap-enabled pods only?

I do not think the swap KEP in it's current shape is GA-ready. If the good use requires to use deprecated flag on kubernetes to disable eviction, it is not a good state.

Reply:
> This will be a big breaking change if we disable eviction once swap is enabled. 

We are not saying that we should disable this. in the code. We want to provide some ways to work with this feature. I think that we can't expect a GA feature to solve all issues. There should be some iteration on more advanced features. Otherwise we will get a lot of features stuck in a permanent beta state.  I think swap is really a feature that users will have to tune to their own purpose. In this case, there is an option to disable memory evictions. Users can also set evictionHard settings to 0 to also avoid memory evictions. Even if we "GA" swap, I still feel that Swap is really meant for advanced users. I don't really know how to fit this into the KEP process. 

> I do not think the swap KEP in it's current shape is GA-ready. If the good use requires to use deprecated flag on kubernetes to disable eviction, it is not a good state.

So this puts the feature in an interesting state. We want to block the GA of Swap until we have a fix for evictions. To probably fix evictions, we are going to need new APIs (at least new eviction signals) so I think we should probably work on a new KEP to unblock this feature.

Some of our usecases have a lot more elaborate requirements for swap eviction ie (eviction on io metrics and psi). In this thread, we are thinking swap eviction based on overall swap totals.

I'm happy to work with others on a new KEP for swap but I think we should be clear on what exactly could unblock this KEP in this current state.

Reply:
In my opinion it is definitely ready for GA. It does not break anything. In the default configuration it will just not allow you to have a RSS larger than physical memory. Just like before. It will still allow swapping and as a result better cache usage etc. Linux can make use of the additional memory.

Then there is a flag to disable eviction (experimental). With that you can choose to allow memory usage above physical memory. You obviously need to be a bit careful as exessive swapping might slow your pods down significantly. In my opinion this could be highlighted a bit more but it is generally explained already.

Reply:
> Some of our usecases have a lot more elaborate requirements for swap eviction ie (eviction on io metrics and psi). In this thread, we are thinking swap eviction based on overall swap totals.

I definitely don't think we should block this feature for advanced eviction that relies on i/o metrics or PSI. 

> I personally see a good argument that we advocate users disable memory eviction if swap is enabled, with the caveat that we intend on improving swappy eviction in the future. How do you feel about that @yujuhong ?

+1 for disabling memory eviction. It seems hard and tricky to set the threshold, and the result may not be as expected. 

For swap-aware eviction, is it possible to consider only the basic case? I think that might still add values to straightforward use cases.

Reply:
I’m not really sure what basic means in this case.

 Do we evict based on overall swap usage? System-oomd has this option where you give a max percentage of swap used. This protects a node from using too much swap. Eviction manager could be extended to track swap memory and we allow an eviction signal for swap, rank pods based on swap usage and delete in that order until swap utilization falls below.
 
For the case of increased memory density, maybe we ignore memory eviction and only evict on swap pressure. Though that would mean that pods that are not swapping could induce memory pressure. So maybe that isn’t a good idea.

Reply:
>> For swap-aware eviction, is it possible to consider only the basic case? I think that might still add values to straightforward use cases.

> I’m not really sure what basic means in this case.

+1.
I feel like the basic case isn't obvious here. Bear in mind that changing how evictions behave needs to be done with extreme care because of backward compatibility reasons. This is one of the reasons I much prefer to raise this discussion in a different KEP and make sure we discuss and understand the different options here.